src/sources/splunk_hec/acknowledgements.rs

```diff
@@ -0,0 +1,467 @@
+                    let expiring_channels = channels
+                        .iter()
+                        .filter_map(|(channel_id, channel)| {
+                            if channel.get_last_used().elapsed().as_secs() >= max_idle_time {
+                                Some(channel_id.clone())
+                            } else {
+                                None
+                            }
+                        })
+                        .collect::<Vec<String>>();
+
+                    for channel_id in expiring_channels {
+                        channels.remove(&channel_id);
+                    }
+                }
+            });
+        }
+
+        Self {
+            max_pending_acks: u64::from(config.max_pending_acks),
+            max_pending_acks_per_channel: u64::from(config.max_pending_acks_per_channel),
+            max_number_of_ack_channels: u64::from(config.max_number_of_ack_channels),
+            channels,
+            shutdown,
+            total_pending_acks: AtomicU64::new(0),
+        }
+    }
+
+    /// Creates a channel with the specified id if it does not exist.
+    async fn create_or_get_channel(&self, id: String) -> Result<Arc<Channel>, Rejection> {
+        let channels = self.channels.read().await;
+        if let Some(channel) = channels.get(&id) {
+            return Ok(Arc::clone(channel));
+        }
+        drop(channels);
+
+        let mut channels = self.channels.write().await;
+        if channels.len() < self.max_number_of_ack_channels as usize {
+            // Create the channel if it does not exist
+            let channel = Arc::new(Channel::new(
+                self.max_pending_acks_per_channel,
+                self.shutdown.clone(),
+            ));
+            channels.insert(id, Arc::clone(&channel));
+            Ok(channel)
+        } else {
+            Err(Rejection::from(ApiError::ServiceUnavailable))
+        }
+    }
+
+    /// Gets the next available ack id from a specified channel, creating the channel if it does not exist
+    pub async fn get_ack_id_from_channel(
+        &self,
+        channel_id: String,
+        batch_rx: BatchStatusReceiver,
+    ) -> Result<u64, Rejection> {
+        let channel = self.create_or_get_channel(channel_id).await?;
+        let ack_id = channel.get_ack_id(batch_rx);
+        let total_pending_acks = self.total_pending_acks.fetch_add(1, Ordering::Relaxed) + 1;
+        if total_pending_acks > self.max_pending_acks {
+            self.total_pending_acks.fetch_sub(
+                self.drop_oldest_pending_ack_per_channel().await,
+                Ordering::Relaxed,
+            );
+        }
+        Ok(ack_id)
+    }
+
+    /// Gets the requested ack id statuses from a specified channel, creating the channel if it does not exist
+    pub async fn get_acks_status_from_channel(
+        &self,
+        channel_id: String,
+        ack_ids: &[u64],
+    ) -> Result<HashMap<u64, bool>, Rejection> {
```

Comment:
Since the incoming `ack_ids` are ordered, could this simply return a `Vec<bool>` indexed by the same?

Reply:
We could return a `Vec<bool>` here and instead create the `HashMap` in the caller; the `HashMap` being necessary for turning into the HTTP response body. Is there an advantage to doing so?

Reply:
Okay, if the caller needs a `HashMap` anyways, then there is no advantage.